File: src/services/currencyService.js
import axios from 'axios';

const API_URL = process.env.REACT_APP_API_URL || 'http://localhost:8080';

/**
 * Currency Service
 * - 재화 관련 API 호출
 */
const currencyService = {
  /**
   * 현재 사용자의 재화 정보 조회
   */
  getCurrency: async () => {
    const token = localStorage.getItem('token');
    if (!token) {
      throw new Error('No authentication token found');
    }

    const response = await axios.get(`${API_URL}/api/currency`, {
      headers: {
        'Authorization': `Bearer ${token}`
      }
    });

    return response.data; // { silverCoins: number, goldCoins: number }
  },

  /**
   * Silver Coin 추가
   */
  addSilverCoins: async (amount) => {
    const token = localStorage.getItem('token');
    if (!token) {
      throw new Error('No authentication token found');
    }

    const response = await axios.post(`${API_URL}/api/currency/silver/add`,
      { amount },
      {
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json'
        }
      }
    );

    return response.data; // { silverCoins: number, goldCoins: number }
  },

  /**
   * Gold Coin 추가
   */
  addGoldCoins: async (amount) => {
    const token = localStorage.getItem('token');
    if (!token) {
      throw new Error('No authentication token found');
    }

    const response = await axios.post(`${API_URL}/api/currency/gold/add`,
      { amount },
      {
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json'
        }
      }
    );

    return response.data; // { silverCoins: number, goldCoins: number }
  },

  /**
   * Silver Coin 차감
   */
  subtractSilverCoins: async (amount) => {
    const token = localStorage.getItem('token');
    if (!token) {
      throw new Error('No authentication token found');
    }

    const response = await axios.post(`${API_URL}/api/currency/silver/subtract`,
      { amount },
      {
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json'
        }
      }
    );

    return response.data; // { silverCoins: number, goldCoins: number }
  },

  /**
   * Gold Coin 차감
   */
  subtractGoldCoins: async (amount) => {
    const token = localStorage.getItem('token');
    if (!token) {
      throw new Error('No authentication token found');
    }

    const response = await axios.post(`${API_URL}/api/currency/gold/subtract`,
      { amount },
      {
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json'
        }
      }
    );

    return response.data; // { silverCoins: number, goldCoins: number }
  },
  /**
   * Gold Coin을 Silver Coin으로 교환 (1:100)
   */
  exchangeGoldToSilver: async (goldAmount) => {
    const token = localStorage.getItem('token');
    if (!token) {
      throw new Error('No authentication token found');
    }

    const response = await axios.post(`${API_URL}/api/currency/exchange/gold-to-silver`,
      { goldAmount },
      {
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json'
        }
      }
    );

    return response.data; // { silverCoins: number, goldCoins: number }
  }
};

export default currencyService;

File: backend/src/main/java/com/community/service/CurrencyService.java
package com.community.service;

import com.community.model.User;
import com.community.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashMap;
import java.util.Map;

/**
 * CurrencyService
 * - 사용자 재화 관리 서비스
 * - Silver Coin (일반 재화), Gold Coin (유료 재화) 처리
 */
@Service
@Slf4j
@RequiredArgsConstructor
public class CurrencyService {

    private final UserRepository userRepository;

    /**
     * 사용자의 재화 정보 조회
     */
    public Map<String, Integer> getUserCurrency(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        Map<String, Integer> currency = new HashMap<>();
        currency.put("silverCoins", user.getSilverCoins());
        currency.put("goldCoins", user.getGoldCoins());
        return currency;
    }

    /**
     * Silver Coin 추가
     */
    @Transactional
    public Map<String, Integer> addSilverCoins(Long userId, Integer amount) {
        if (amount <= 0) {
            throw new IllegalArgumentException("Amount must be positive");
        }

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        user.setSilverCoins(user.getSilverCoins() + amount);
        userRepository.save(user);

        Map<String, Integer> currency = new HashMap<>();
        currency.put("silverCoins", user.getSilverCoins());
        currency.put("goldCoins", user.getGoldCoins());
        return currency;
    }

    /**
     * Gold Coin 추가
     */
    @Transactional
    public Map<String, Integer> addGoldCoins(Long userId, Integer amount) {
        if (amount <= 0) {
            throw new IllegalArgumentException("Amount must be positive");
        }

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        user.setGoldCoins(user.getGoldCoins() + amount);
        userRepository.save(user);

        Map<String, Integer> currency = new HashMap<>();
        currency.put("silverCoins", user.getSilverCoins());
        currency.put("goldCoins", user.getGoldCoins());
        return currency;
    }

    /**
     * Silver Coin 차감
     */
    @Transactional
    public Map<String, Integer> subtractSilverCoins(Long userId, Integer amount) {
        if (amount <= 0) {
            throw new IllegalArgumentException("Amount must be positive");
        }

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (user.getSilverCoins() < amount) {
            throw new IllegalArgumentException("Insufficient silver coins");
        }

        user.setSilverCoins(user.getSilverCoins() - amount);
        userRepository.save(user);

        Map<String, Integer> currency = new HashMap<>();
        currency.put("silverCoins", user.getSilverCoins());
        currency.put("goldCoins", user.getGoldCoins());
        return currency;
    }

    /**
     * Gold Coin 차감
     */
    @Transactional
    public Map<String, Integer> subtractGoldCoins(Long userId, Integer amount) {
        if (amount <= 0) {
            throw new IllegalArgumentException("Amount must be positive");
        }

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (user.getGoldCoins() < amount) {
            throw new IllegalArgumentException("Insufficient gold coins");
        }

        user.setGoldCoins(user.getGoldCoins() - amount);
        userRepository.save(user);

        Map<String, Integer> currency = new HashMap<>();
        currency.put("silverCoins", user.getSilverCoins());
        currency.put("goldCoins", user.getGoldCoins());
        return currency;
    }
    /**
     * Gold Coin을 Silver Coin으로 교환 (1 Gold = 100 Silver)
     */
    @Transactional
    public Map<String, Integer> exchangeGoldToSilver(Long userId, Integer goldAmount) {
        if (goldAmount <= 0) {
            throw new IllegalArgumentException("Amount must be positive");
        }

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (user.getGoldCoins() < goldAmount) {
            throw new IllegalArgumentException("Insufficient gold coins for exchange");
        }

        // 금화 차감 및 은화 추가 (1:100 비율)
        int silverToAdd = goldAmount * 100;
        user.setGoldCoins(user.getGoldCoins() - goldAmount);
        user.setSilverCoins(user.getSilverCoins() + silverToAdd);
        
        userRepository.save(user);

        log.info("[CurrencyService] 재화 교환 완료: userId={}, goldSpent={}, silverGained={}", 
            userId, goldAmount, silverToAdd);

        Map<String, Integer> currency = new HashMap<>();
        currency.put("silverCoins", user.getSilverCoins());
        currency.put("goldCoins", user.getGoldCoins());
        return currency;
    }
}
